domain/application/state/application.go

```diff
@@ -575,8 +558,166 @@ ON CONFLICT(net_node_uuid) DO UPDATE
+	// net node corresponding to the cloud container.
+	cloudContainerDeviceInfo := cloudContainerDevice{
+		Name:              address.Device.Name,
+		NetNodeID:         netNodeID,
+		DeviceTypeID:      int(address.Device.DeviceTypeID),
+		VirtualPortTypeID: int(address.Device.VirtualPortTypeID),
+	}
+	if cloudContainerDeviceInfo.Name == "" {
+		return fmt.Errorf("cloud container placeholder device must have a name supplied")
+	}
+
+	selectCloudContainerDeviceStmt, err := st.Prepare(`
+SELECT &cloudContainerDevice.uuid FROM link_layer_device
+WHERE net_node_uuid = $cloudContainerDevice.net_node_uuid
+`, cloudContainerDeviceInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertCloudContainerDeviceStmt, err := st.Prepare(`
+INSERT INTO link_layer_device (*) VALUES ($cloudContainerDevice.*)
+`, cloudContainerDeviceInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// See if the link layer device exists, if not insert it.
+	var deviceID string
+	err = tx.Query(ctx, selectCloudContainerDeviceStmt, cloudContainerDeviceInfo).Get(&cloudContainerDeviceInfo)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		deviceUUID, err := uuid.NewUUID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		cloudContainerDeviceInfo.UUID = deviceUUID.String()
+		if err := tx.Query(ctx, insertCloudContainerDeviceStmt, cloudContainerDeviceInfo).Run(); err != nil {
+			return errors.Annotatef(err, "inserting cloud container device for unit %q", unitName)
+		}
+	} else if err != nil {
+		return errors.Annotatef(err, "querying cloud container link layer device for unit %q", unitName)
+	}
+	deviceID = cloudContainerDeviceInfo.UUID
+
+	// Now process the address details.
+	ipAddr := ipAddress{
+		Value:        address.Value,
+		ConfigTypeID: int(address.ConfigType),
+		TypeID:       int(address.AddressType),
+		OriginID:     int(address.Origin),
+		ScopeID:      int(address.Scope),
+		DeviceID:     deviceID,
+	}
+
+	selectAddressUUIDStmt, err := st.Prepare(`
+SELECT &ipAddress.uuid
+FROM ip_address
+WHERE device_uuid = $ipAddress.device_uuid;
+`, ipAddr)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	upsertAddressStmt, err := sqlair.Prepare(`
+INSERT INTO ip_address (*)
+VALUES ($ipAddress.*)
+ON CONFLICT(uuid) DO UPDATE SET
+    address_value = excluded.address_value,
+    type_id = excluded.type_id,
+    scope_id = excluded.scope_id,
+    origin_id = excluded.origin_id,
+    config_type_id = excluded.config_type_id
+`, ipAddr)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Container addresses are never deleted unless the container itself is deleted.
+	// First see if there's an existing address recorded.
+	err = tx.Query(ctx, selectAddressUUIDStmt, ipAddr).Get(&ipAddr)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return fmt.Errorf("querying existing cloud container address for device %q: %w", deviceID, err)
+	}
+
+	// Create a UUID for new addresses.
+	newAddr := ipAddr.AddressUUID == ""
```

Comment:
Poorly named. `newAddr` sounds like this is a new address that you have made not a flag to tell us if we should make a new address uuid.

In any event uuid generation should be pushed upwards like the previous comments.

Reply:
Fixed name